networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/AttestationValidator.java

```diff
@@ -0,0 +1,212 @@
+  private static final UnsignedLong MAXIMUM_GOSSIP_CLOCK_DISPARITY =
+      UnsignedLong.valueOf(Constants.MAXIMUM_GOSSIP_CLOCK_DISPARITY);
+
+  private final Set<ValidatorAndSlot> receivedValidAttestations =
+      ConcurrentLimitedSet.create(VALID_ATTESTATION_SET_SIZE, LimitStrategy.DROP_OLDEST_ELEMENT);
+  private final RecentChainData recentChainData;
+
+  public AttestationValidator(final RecentChainData recentChainData) {
+    this.recentChainData = recentChainData;
+  }
+
+  public ValidationResult validate(
+      final Attestation attestation, final UnsignedLong receivedOnSubnetId) {
+    ValidationResult validationResult = singleAttestationChecks(attestation, receivedOnSubnetId);
+    if (validationResult != VALID) {
+      return validationResult;
+    }
+
+    validationResult = singleOrAggregateAttestationChecks(attestation);
+    if (validationResult != VALID) {
+      return validationResult;
+    }
+
+    return firstValidAttestationCheck(attestation);
+  }
+
+  private ValidationResult firstValidAttestationCheck(final Attestation attestation) {
+    // The attestation is the first valid attestation received for the participating validator for
+    // the slot, attestation.data.slot.
+    if (!receivedValidAttestations.add(getValidatorAndSlot(attestation))) {
+      return INVALID;
+    }
+    return VALID;
+  }
+
+  private ValidationResult singleAttestationChecks(
+      final Attestation attestation, final UnsignedLong receivedOnSubnetId) {
+    // The attestation's committee index (attestation.data.index) is for the correct subnet.
+    if (!CommitteeUtil.getSubnetId(attestation).equals(receivedOnSubnetId)) {
+      return INVALID;
+    }
+
+    // The attestation is unaggregated -- that is, it has exactly one participating validator
+    // (len([bit for bit in attestation.aggregation_bits if bit == 0b1]) == 1).
+    if (attestation.getAggregation_bits().countSetBits() != 1) {
+      return INVALID;
+    }
+
+    // The attestation is the first valid attestation received for the participating validator for
+    // the slot, attestation.data.slot.
+    if (receivedValidAttestations.contains(getValidatorAndSlot(attestation))) {
+      return INVALID;
+    }
+    return VALID;
+  }
+
+  private ValidationResult singleOrAggregateAttestationChecks(final Attestation attestation) {
+    // attestation.data.slot is within the last ATTESTATION_PROPAGATION_SLOT_RANGE slots (within a
+    // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance) -- i.e. attestation.data.slot +
+    // ATTESTATION_PROPAGATION_SLOT_RANGE >= current_slot >= attestation.data.slot (a client MAY
+    // queue
+    // future attestations for processing at the appropriate slot).
+    final UnsignedLong currentTimeMillis = secondsToMillis(recentChainData.getStore().getTime());
+    if (isAfterPropagationSlotRange(currentTimeMillis, attestation)) {
+      return INVALID;
+    }
+    if (isBeforeMinimumBroadcastTime(attestation, currentTimeMillis) < 0) {
```

Comment:
nit: it's better to have these two functions match each other's style
```suggestion
    if (isAfterPropagationSlotRange(currentTimeMillis, attestation)) {
      return INVALID;
    }
    if (isBeforeMinimumBroadcastTime(attestation, currentTimeMillis)) {
```

Reply:
Good spot.